HW0/P5/P5.txt

```diff
@@ -1 +1,41 @@
+(256/128−1) + log_2(128) = 8 seconds
+
+4. If you were feeling especially frustrated, perhaps you may use more than 256 bags. Instead, assume there are N bags. How long will it take the infinite number of employees to count the total? Using Python, plot the result with bag count on the horizontal and counting time on the vertical. On the same graph, plot the time it would take the lone cashier to do this by herself.
+
+See P4.png
+
+5. We’ve been neglecting any overhead associated with people communicating. How does the answer for 256 bags change, if it takes one second to communicate a number between two employees?
+
+In order to account for the communication between employees we must re-examine the counting process. Communication will only need to take place when employees have added all of their coin bags and need to sum with the employee next to them. This means that our original algorithm still holds true (B/E)-1 but now we must account for the employees communicating. This can once again be modeled with the logarithm, log_2(E). So to communicate and add between employees, the algorithm is now log_2(E) + log_2(E) or 2 * log_2(E). Putting it all together, I have the following: (B/E)-1 + 2 * log_2(E) and in our problem this equates to 256/8 - 1 + 2 * log_2(8) = 32-1+2*3 = 37 seconds. In the case of having infinite employees or our maximum needed of 128 we arrive at 256/128 -1 + 2*log_2(7) = 2-1+2*7 = 15 seconds. So the communication between employees adds some additional time but it doesn't seem like a huge bottleneck for this amount of bags. If they had to count 1 million bags, this communication time could start to add quite a bit of additional time.
+
+6. Also assume it takes one second to hand a bag (or any number of bags) between two people (including from you to a cashier). How long will it take to verify 256 bags worth of coins if each cashier comes to you and takes 1 bag? What about 2 bags? Is there a better strategy? Assume you can’t be handing bags to more than one person at a time.
+
+In order to verify the 256 bags worth coins, given that it takes exactly 1 second to hand a bag of coins to a cashier we need to examine how long it takes to hand all of the bags to the employees. In the case of 8 employees working, it will take exactly 8 seconds to hand over the first round of bags, then an additional 8 seconds to hand of the next round and so on and so forth. The trick here is that one employee can start summing their bags as you hand of bags to the next employee. So to hand over 2 rounds of bags, and to allow time for all employees to add their bags, it will take 8+8+1=17 seconds. If we apply this logic to all bags, it takes 17 seconds for 2 rounds or 16 bags, it will take 16 rounds of 16 bags to complete the task or 16*16+1 = 257. Now for the communication costs (log_2(E)) and the additional time needed to add (log_2(E)) we add the 2 * log_2(E) factor and arrive at 257 + 2 * log_2(8) = 256+(2*3) = 263.
+
```

Comment:
Interesting, now that I read through the problem again and think about it, it seems that you are right in that it will take more than 6 additional seconds for the last employee to add his/her bags. I think your answer may be correct, I hope that they release the solutions soon!